File: src/main/java/no/fintlabs/integration/validation/constraints/HibernateConstraintValidator.java
package no.fintlabs.integration.validation.constraints;

import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.lang.annotation.Annotation;

public interface HibernateConstraintValidator<A extends Annotation, T> extends ConstraintValidator<A, T> {

    @Override
    default boolean isValid(T value, ConstraintValidatorContext constraintValidatorContext) {
        if (constraintValidatorContext instanceof HibernateConstraintValidatorContext) {
            HibernateConstraintValidatorContext hibernateConstraintValidatorContext =
                    constraintValidatorContext.unwrap(HibernateConstraintValidatorContext.class);
            return isValid(value, hibernateConstraintValidatorContext);
        }
        throw new IllegalStateException("Validator is not HibernateConstraintValidatorContext");
    }

    boolean isValid(T value, HibernateConstraintValidatorContext hibernateConstraintValidatorContext);

}

File: src/main/kotlin/no/novari/flyt/integration/validation/constraints/HibernateConstraintValidator.kt
package no.novari.flyt.integration.validation.constraints

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext

interface HibernateConstraintValidator<A : Annotation, T> : ConstraintValidator<A, T> {
    override fun isValid(
        value: T?,
        constraintValidatorContext: ConstraintValidatorContext,
    ): Boolean {
        if (constraintValidatorContext is HibernateConstraintValidatorContext) {
            val hibernateConstraintValidatorContext =
                constraintValidatorContext.unwrap(HibernateConstraintValidatorContext::class.java)
            return isValid(value, hibernateConstraintValidatorContext)
        }

        throw IllegalStateException("Validator is not HibernateConstraintValidatorContext")
    }

    fun isValid(
        value: T?,
        hibernateConstraintValidatorContext: HibernateConstraintValidatorContext,
    ): Boolean
}
